src/range.cr

```diff
@@ -252,6 +252,56 @@ struct Range(B, E)
     {% end %}
   end
 
+  # Returns the union of this range, and another.
+  # Returns 0..0 if there is no overlap.
+  #
+  # ```
+  # (1..5).union(5..10)   # => 1..10
+  # (1...10).union(1..10) # => 1..10
+  # ```
+  def union(other : Range)
+    e_s : E = self.end
+    e_o : E = other.end
+
+    return if self.end < other.begin && (!e_s.responds_to?(:succ) || e_s.succ != other.begin)
+    return if other.end < self.begin && (!e_o.responds_to?(:succ) || e_o.succ != self.begin)
+
+    larger = (self.end > other.end)
+    Range.new(
+      self.begin < other.begin ? self.begin : other.begin,
+      self.end > other.end ? self.end : other.end,
+      exclusive: (larger && self.excludes_end?) || (!larger && other.excludes_end?),
+    )
+  end
+
+  # Returns the intersection of this range, and another.
+  # Returns 0..0 if there is no overlap.
+  #
+  # ```
+  # (2..10).intersection(0..8)   # => 2..8
+  # (1...10).intersection(7..12) # => 7...10
+  # ```
+  def intersection(other : Range)
+    return if self.end < other.begin || other.end < self.begin
+
+    larger = (self.end > other.end)
+    Range.new(
+      self.begin > other.begin ? self.begin : other.begin,
+      self.end < other.end ? self.end : other.end,
+      exclusive:(!larger && self.excludes_end?) || (larger && other.excludes_end?),
+    )
```

Comment:
**issue:** We're completely missing tests for `exclusive` ranges.
And it looks like the current implementation might actually not be correct. It seems to take `excludes_end?` of ther smaller of `self` and `other`, but shouldn't it be the larger of both?

I believe this could be implemented in a way that's simpler and more straightforward to reason about:

```suggestion
    larger = (self.end > other.end)
    Range.new(
      Math.max(self.begin, other.begin),
      Math.min(self.end, other.end),
      exclusive:({self, other}.max_by(&.end).excludes_end?,
    )
```

Ditto for `#union`.